File: app/src/main/java/ru/arhlib/app/news/PostAdapter.java
package ru.arhlib.app.news;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;

import ru.arhlib.app.actual.ActualItemViewHolder;
import ru.arhlib.app.databinding.PostItemBinding;

public class PostAdapter extends ListAdapter<Post, PostAdapter.PostViewHolder> {

    public PostAdapter() {
        super(new PostDiffCallback());
    }

    @NonNull
    @Override
    public PostViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        PostItemBinding binding = PostItemBinding.inflate(LayoutInflater.from(parent.getContext()),
                parent, false);
        return new PostViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull PostViewHolder holder, int position) {
        holder.postBinding.setCallback(new PostCallback());
        holder.postBinding.setPost(getItem(position));
        holder.postBinding.executePendingBindings();
    }

    class PostViewHolder extends ActualItemViewHolder {
        PostViewHolder(PostItemBinding binding) {
            super(binding);
        }
    }
}

File: app/src/main/java/ru/arhlib/app/news/PostAdapter.kt
package ru.arhlib.app.news

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import ru.arhlib.app.databinding.PostItemBinding

class PostAdapter : PagingDataAdapter<Post, PostViewHolder>(PostDiffCallback) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PostViewHolder {
        val binding = PostItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return PostViewHolder(binding)
    }

    override fun onBindViewHolder(holder: PostViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    private object PostDiffCallback : DiffUtil.ItemCallback<Post>() {
        override fun areItemsTheSame(oldPost: Post, newPost: Post): Boolean {
            return oldPost.id == newPost.id
        }

        override fun areContentsTheSame(oldPost: Post, newPost: Post): Boolean {
            return oldPost == newPost
        }
    }
}
